docs/content/sips/0xx-component-dependencies.md

```diff
@@ -0,0 +1,108 @@
+title = "SIP 0XX - Component Dependencies"
+template = "main"
+date = "2024-05-31T12:00:00Z"
+---
+Summary: This feature enables developers to specify and manage dependencies for Spin components, allowing the use of components from various languages and sources, either locally or remotely, with configurable inheritance of parent component configuration.
+
+Owner(s): [brian.hardock@fermyon.com](mailto:brian.hardock@fermyon.com)
+
+Created: May 31, 2024
+
+# Background
+This feature enables the polyglot re-use of components, allowing developers to specify how to use components written in various languages as libraries to fulfill dependencies in their Spin components.
+
+For instance, consider writing a Spin component in Go that depends on the `aws:client/s3` interface (expressed as an import in the compiled Go component, e.g., `import aws:client/s3`). To satisfy this dependency, you can use a componentized S3 client written in Rust that exports an instance of the `aws:client/s3` interface. Using the proposed syntax in this SIP, Spin developers can instruct Spin on how to compose these components together to achieve this.
+
+# Proposal
+Developers can use the following syntax to specify that they want to use a component as a dependency, either from a local disk or via a remote component registry:
+
+## Using dependencies from disk (local)
+```toml
+[component."infra-dashboard".dependencies]
+# Components can satisfy dependencies by import name (e.g., "aws:client/s3")
+# using a component from disk. It is implicitly assumed that the
+# `my_aws_client.wasm` component exports an instance of the
+# `aws:client/s3` interface.
+"aws:client/s3" = { path = "my_aws_client.wasm" }
+
+# Optionally, explicitly specify the export name to use to satisfy the `aws:client/s3` interface.
+"aws:client/s3" = { path = "my_aws_client.wasm", export = "my-s3-client" }
+
+# Without an interface name, attempt to resolve every import of the "aws:client" package.
```

Comment:
Does this include wasi imports? What happens if the provided wasm file does not include an export that satisfies an import? Do we error early or do we simply implicitly import anything that was not satisified?

Reply:
> Does this include wasi imports?

I don't have a strong technical argument for why to exclude wasi imports but i'm also tending towards being a little conservative here in the initial MVP. So maybe we should error initially if a wasi imports is listed in dependencies? We can always relax this constraint in the future.

> What happens if the provided wasm file does not include an export that satisfies an import?

This is a good question that i will update the SIP to answer once we decide what should be the expected behavior here. My current thinking (for the MVP) is that we should tend towards simply importing anything that was not satisfied.

I would like to reject and error on invalid dependencies. So if a dependency points to component that can't satisfy anything that should probably error.